core/src/term/array.rs

```diff
@@ -157,29 +157,50 @@ impl IntoIterator for Array {
         // to the inner slice, then we can:
         // - Drop the elements outside our inner view,
         // - Move out the elements inside out inner view.
+        // - Drop the rest of the elements when we're dropped
         // Otherwise, we clone everything.
-
-        unsafe {
-            let mut inner: Rc<[ManuallyDrop<RichTerm>]> = transmute(self.inner);
-            let idx = self.start;
-            let end = self.end;
-
-            if let Some(slice) = Rc::get_mut(&mut inner) {
-                for term in &mut slice[..idx] {
+        //
+        // If we start as a shared reference, we could become the only reference
+        // later, but we clone everything anyways, so we don't end up in an
+        // invalid in-between state where some terms have been freed manually
+        // and others haven't.
```

Comment:
Yeah, I was saying something else. In retrospect this comment wasn't super clear. This wasn't meant to explain why this code is safe, it's meant to explain why we don't optimize further. In theory if we become the only reference while we're iterating, we could at that point not bother cloning things. This is what the code did before this change. But because this could leave us in an invalid state that is hard to clean up, I think we shouldn't bother with that optimization.

I'll try to reword the comment to be clearer.

And yes, you're right that we don't clone anything at this point in the code, we just mark it as "to be cloned".